src/sentry/workflow_engine/migrations/0054_clean_up_orphaned_metric_alert_objects.py

```diff
@@ -0,0 +1,76 @@
+# Generated by Django 5.1.7 on 2025-05-01 20:38
+
+import logging
+
+from django.db import migrations, router, transaction
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Exists, OuterRef
+
+from sentry.new_migrations.migrations import CheckedMigration
+
+logger = logging.getLogger(__name__)
+
+
+def delete_orphaned_migrated_metric_alert_objects(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    Action = apps.get_model("workflow_engine", "Action")
+    DataConditionGroup = apps.get_model("workflow_engine", "DataConditionGroup")
+    DataConditionGroupAction = apps.get_model("workflow_engine", "DataConditionGroupAction")
+    Detector = apps.get_model("workflow_engine", "Detector")
+    Workflow = apps.get_model("workflow_engine", "Workflow")
+    WorkflowDataConditionGroup = apps.get_model("workflow_engine", "WorkflowDataConditionGroup")
+
+    orphaned_dcgs = (
+        DataConditionGroup.objects.filter(
+            ~Exists(Detector.objects.filter(workflow_condition_group_id=OuterRef("id")))
+        )
+        .filter(
+            ~Exists(WorkflowDataConditionGroup.objects.filter(condition_group_id=OuterRef("id")))
+        )
+        .filter(~Exists(Workflow.objects.filter(when_condition_group_id=OuterRef("id"))))
+    )
+
+    orphaned_action_ids = DataConditionGroupAction.objects.filter(
+        Exists(orphaned_dcgs.filter(id=OuterRef("condition_group_id")))
+    ).values_list("action__id", flat=True)
+
+    orphaned_actions = Action.objects.filter(id__in=orphaned_action_ids)
+
+    logger.info("orphaned action count: %s", orphaned_actions.count())
+    logger.info("orphaned dcg count: %s", orphaned_dcgs.count())
+
+    with transaction.atomic(router.db_for_write(Action)):
+        for action in orphaned_actions:
+            action.delete()
+        for dcg in orphaned_dcgs:
+            dcg.delete()
```

Comment:
I'd also recommend not wrapping this entire operation in a transaction, because all involved rows will be locked for the whole transaction. Maybe it doesn't matter too much since they're orphaned, but better to be safe.

If you want to make sure the rows related to each other are in a transaction that's fine, but you'll just need to associate them and have a smaller transaction for each set of related rows. I think it's probably safe enough to just skip it though.